Add GitHub issue forms for bug reports and feature requests

TechnoPorg opened the pull request:
Closes #6823

Issue forms are technically in "beta", but they've been in beta for 3 years, so I think it's safe to say that they're not going anywhere.

The use of issue forms provides a cleaner experience both for users and contributors:
- Users receive more guidance on the specifics of submitting an issue thanks to the ability for contributors to provide additional context in the form.
- It provides a more consistent, centralized method for contributors to ensure that all issues contain the necessary information in a cleanly structured format.

I tried to strike a balance between keeping the content of the existing issue template, while also providing a few more specifics to ensure higher quality bug reports. Some general inspiration was taken from Godot's issue templates.

TechnoPorg commented:
It can be tested out on [my fork](https://github.com/TechnoPorg/lmms).

What the forms look like in action:

![image](https://github.com/LMMS/lmms/assets/69441745/4d9954c4-d4d3-4193-80ff-3ec87e9d161f)
----------------------------------------------------------------------------------------------------------------------------------------------------------
![image](https://github.com/LMMS/lmms/assets/69441745/ecb3a50d-7f20-4b42-a1af-207d81abed68)

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,85 @@
+      label: Most Recent Working Version
+      description: |
+        - If there is a previous version of LMMS that did not exhibit the bug, please include it here.
+        - Using `git bisect` to determine exactly where the regression occurred will make it easier for LMMS developers to find and fix the bug.
```
I worry mentioning `git bisect` could confuse or scare off less technical users. Those who know how to use it probably know to do so anyway. Also, it reports the first bad commit, not the last good one, so we risk people putting the wrong version in this field.

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,85 @@
+      label: Logs
+      description: |
+        - Please copy and paste any relevant log output where instructed.
+        - LMMS log output can be obtained by running it via the console.
```
LMMS doesn't produce console output by default on Windows. Also, "where instructed" doesn't make sense to me here - are we guiding users through this form elsewhere? I would have thought this form itself would usually constitute the instructions.

TechnoPorg replied:
I had been referring to the comment inside the HTML block when I said "where instructed", but I think that it's actually clear enough by itself and what I said doesn't make sense, you're right.

I took out the note about console output entirely and left it to the user, but I'm not sure if that's the right approach either.

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,85 @@
+      label: Screenshots / Minimum Reproducible Project (MRP)
+      description: |
+        - Please upload any screenshots showing the bug in action.
+        - If possible, also include a .mmp/.mmpz project containing the simplest possible setup needed to reproduce the bug.
```
GitHub only allows a files with a limited set of extensions to be uploaded. I think it would be worth adding that the extension needs to be changed first.

TechnoPorg replied:
I made a note that files should be zipped first, which I think might be more intuitive for the average user.

DomClark commented on .github/ISSUE_TEMPLATE/feature_request.yml:
```diff
@@ -0,0 +1,35 @@
+      label: Enhancement Summary
+      description: |
+        - Briefly describe the enhancement.
+        - Explain how the proposed enhancement helps overcome a limitation of LMMS you are currently facing.
```
How is this second point different from the "justification" section that comes next?

TechnoPorg replied:
On second thought, it's not, so I merged both sections together.

DomClark commented on .github/ISSUE_TEMPLATE/feature_request.yml:
```diff
@@ -0,0 +1,35 @@
+      label: Justification
+      description: |
+        - Describe why you believe this to be a good idea, and how you think it would improve LMMS.
+        - Is there a reason why this should be included in LMMS itself and not as a plugin?
```
We don't have a policy of not implementing things that are available as third-party plugins (except perhaps instruments that include big sample libraries), so I'm not sure if this point is necessary.

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,85 @@
+body:
+  - type: markdown
+    attributes:
+      value: "# Please search the issue tracker for existing bug reports before submitting your own."
```
Would this be better as a checkbox at the end of the form that the user has to tick before they can submit it?

TechnoPorg replied:
I hadn't thought of that, but you're right, that's a good idea.

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,85 @@
+    attributes:
+      label: Steps To Reproduce
+      description: |
+        - Describe the minimum steps one can follow to reproduce this bug for themself.
```
The reflexive form of "one" is "oneself", but I think this may be better without any reference to the person performing the steps, e.g. "describe the minimum set of steps required to reproduce this bug".

TechnoPorg commented:
@DomClark Thanks for taking the time to give such a thorough review! I've made some changes that should address most of your concerns, so if you could give it a second look that would be great.

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,89 @@
+      label: Steps To Reproduce
+      description: |
+        - Describe the minimum set of steps required to reproduce this bug.
+        - If you included an MRP below, you can describe here how it should be used.
```
I think it would be worth defining "MRP" at its first use (or even avoiding the acronym altogether, for maximum clarity).

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,89 @@
+        - Please upload any screenshots showing the bug in action.
+        - If possible, also include a .mmp/.mmpz project containing the simplest possible setup needed to reproduce the bug.
+
+        ***Note:** To upload a project file to GitHub, it will need to placed in a .zip archive.*
```
```suggestion
        ***Note:** To upload a project file to GitHub, it will need to be placed in a .zip archive.*
```

DomClark commented on .github/ISSUE_TEMPLATE/feature_request.yml:
```diff
@@ -0,0 +1,33 @@
+      label: Implementation Details / Mockup
+      description: |
+        - Explain how you believe this enhancement should be implemented.
+        - If your proposal encompasses changes to the user interface, please include **diagrams** displaying your intent.
```
I'm not sure whether "diagrams" needs to be in bold here. No other text besides "Note:" in the other form is bold, and it seems unambiguous enough on its own.

DomClark commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,89 @@
+    id: expected-behaviour
+    attributes:
+      label: Expected Behaviour
+      description: Describe what should have happened and how LMMS should have behaved.
```
I think "what should have happened" and "how LMMS should have behaved" are more or less the same thing. The previous description has a similar issue too.

DomClark reviewed (commented):
Sorry for being particularly picky with this - it's going to be viewed by people who may be less skilled with technology, or English, so I'm trying to polish it a bit.

One other thing I'd point out, that I haven't put in a separate comment on a file, is that the style of instruction varies between some of the steps. Sometimes the imperative alone is used, and sometimes "please" is included too. I would suggest omitting "please" for all fields that will appear in the issue itself (as these are instructions on how to fill out the form), but continuing to use it for the final checkbox (as this is a request from us to try to avoid duplicates).

DomClark commented:
One more thing - do these files pass `yamllint`? Johannes is introducing a `.yamllint` configuration file in #6983, which you could use for testing. It specifies a maximum line length of 120 characters, which I suspect may be a bit short for these files, so you may need to disable the line length rule at appropriate places (perhaps even for the whole file if there are enough places to make the linter comments spammy).

TechnoPorg commented:
> Sorry for being particularly picky with this - it's going to be viewed by people who may be less skilled with technology, or English, so I'm trying to polish it a bit.

No problem at all - I actually really appreciate it, so thanks! We both want this PR to be as good as possible, because it deals with one of the most important parts of the average user's experience.

> One more thing - do these files pass yamllint?

I added a yamllint directive to disable line length checks for the entire file, because there is no way to put comments in a multiline string, so it just can't be done on a per-line basis.

Otherwise, I've addressed all the review comments you left plus the one about imperative instructions, so :crossed_fingers: it's good now. Is there anything not there that you'd like to see added?
